File: pgpainless-core/src/main/java/org/pgpainless/exception/SignatureValidationException.java
// SPDX-FileCopyrightText: 2021 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.exception;

import java.util.Map;

import org.bouncycastle.openpgp.PGPException;
import org.bouncycastle.openpgp.PGPSignature;
import org.pgpainless.algorithm.SignatureType;

public class SignatureValidationException extends PGPException {

    public SignatureValidationException(String message) {
        super(message);
    }

    public SignatureValidationException(String message, Exception underlying) {
        super(message, underlying);
    }

    public SignatureValidationException(String message, Map<PGPSignature, Exception> rejections) {
        super(message + ": " + exceptionMapToString(rejections));
    }

    private static String exceptionMapToString(Map<PGPSignature, Exception> rejections) {
        StringBuilder sb = new StringBuilder();
        sb.append(rejections.size()).append(" rejected signatures:\n");
        for (PGPSignature signature : rejections.keySet()) {
            String typeString;
            SignatureType type = SignatureType.fromCode(signature.getSignatureType());
            if (type == null) {
                typeString = "0x" + Long.toHexString(signature.getSignatureType());
            } else {
                typeString = type.toString();
            }
            sb.append(typeString).append(' ')
                    .append(signature.getCreationTime()).append(": ")
                    .append(rejections.get(signature).getMessage()).append('\n');
        }
        return sb.toString();
    }
}

File: pgpainless-core/src/main/kotlin/org/pgpainless/exception/SignatureValidationException.kt
// SPDX-FileCopyrightText: 2025 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.exception

import org.bouncycastle.openpgp.PGPException
import org.bouncycastle.openpgp.PGPSignature
import org.pgpainless.algorithm.SignatureType

class SignatureValidationException : PGPException {

    constructor(message: String?) : super(message)

    constructor(message: String?, underlying: Exception) : super(message, underlying)

    constructor(
        message: String,
        rejections: Map<PGPSignature, Exception>
    ) : super("$message: ${exceptionMapToString(rejections)}")

    companion object {
        @JvmStatic
        private fun exceptionMapToString(rejections: Map<PGPSignature, Exception>): String =
            buildString {
                append(rejections.size).append(" rejected signatures:\n")
                for (signature in rejections.keys) {
                    append(sigTypeToString(signature.signatureType))
                        .append(' ')
                        .append(signature.creationTime)
                        .append(": ")
                        .append(rejections[signature]!!.message)
                        .append('\n')
                }
            }

        @JvmStatic
        private fun sigTypeToString(type: Int): String =
            SignatureType.fromCode(type)?.toString()
                ?: "0x${java.lang.Long.toHexString(type.toLong())}"
    }
}
